File: C#/0817. Linked List Components.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int NumComponents(ListNode head, int[] G) {
        HashSet<int> GSet=new HashSet<int>(G);
        int cnt=0;
        bool flag=false;
        while(head!=null){
            if(GSet.Contains(head.val) && flag==false){
                cnt++;
                flag=true;   
            }
            else if(!GSet.Contains(head.val)){
                flag=false;
            }
            head=head.next;
        }
        return cnt;
    }
}

File: 20190519python/0817. Linked List Components.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def numComponents(self, head: ListNode, G: List[int]) -> int:
        g_set=set(G)
        cnt=0
        rep=0
        flag=True
        while head:
            if head.val in g_set:
                if flag==True:
                    flag=False
                    cnt+=1
            else:
                if cnt>0:
                    rep+=1
                    flag=True
                    cnt=0
            head=head.next
        if cnt>0:
            rep+=1
        return rep
